src/modes/edit_message.cc

```diff
@@ -719,15 +738,50 @@ namespace Astroid {
       account_no ++;
     }
 
+    update_message_id_rhs_from_sender(a->email);
+    
     bool same_account = (rn == from_combo->get_active_row_number ());
     LOG (debug) << "same account: " << same_account;
     if (!same_account) {
       reset_signature ();
     }
-
+    
     return same_account;
   }
 
+  void EditMessage::update_message_id_rhs_from_sender (ustring _from) {
+    ustring hostname = astroid->config ().get <string> ("mail.message_id_fqdn");
+
+    // if the mail.message_id_fqdn parameter contains "%1", update the
+    // msg_id with the expanded value of mail.message_id_fqdn, using
+    // the FQDN from the new _from address
+    if (   ( ! _from.empty())
+        && (hostname.find ("%1", 0) != std::string::npos)
+        && ( ! msg_id.empty())
+       ) {
+      // replace "%1" in the config string with the FQDN part of the
+      // _from address
+      ustring msg_id_rhs = ustring::compose (hostname,
+                                             boost::replace_all_copy (string(fqdn_of_address(_from)), "%", "%%"));
+
+      // replace the RHS (everything to the right of the "@") of the
+      // msg_id with the expanded config string
+      ustring msg_id_lhs = msg_id.substr(0, msg_id.find("@"));
+      msg_id = ustring::compose ("%1@%2", msg_id_lhs, msg_id_rhs);
+    }
+  }
+
+  ustring EditMessage::fqdn_of_address (ustring _addr) {
+    size_t atpos = _addr.find("@");
+    ustring fqdn;
+
+    if ( ! _addr.empty()) {
+      fqdn = _addr.substr(atpos + 1, _addr.length() - atpos);
+    }
```

Comment:
What happens with invalid e-mail addresses not containing any @ ? Does this work on a local e-mail address? From root to charlie?

Reply:
I'll meditate over this one and think about suitable safeguards.